R/ratio_calculations.R

```diff
@@ -62,3 +66,281 @@ multiple_matrix_ratios <- function(data, matrix_col, conc_col, units_col,
+}
+#' @export
+calculate_ratios.data.frame <- function(
+  data,
+  parameter,
+  match_cols,
+  ref_groups,
+  test_groups = NULL,
+  adjusting_factor = 1,
+  custom_pptestcd = NULL
+) {
+  if (!any(data$PPTESTCD == parameter)) {
+    warning(
+      paste0(
+        "No parameter with PPTESTCD: '",
+        paste(parameter, collapse = ","),
+        "' is not found in the PKNCA results."
+      )
+    )
+  }
+
+  # Deduce what are all the group columns in the result data.frame
+  extra_res_cols <- c(
+    "PPTEST", "PPORRES", "PPORRESU", "PPSTRES", "PPSTRESU", "type_interval", "exclude"
+  )
+  group_cols <- setdiff(colnames(data), extra_res_cols)
+
+  # Filter for the parameter of interest
+  df <- data[data$PPTESTCD == parameter, ]
+
+  # Define the denominator rows
+  df_den <- merge(df, ref_groups)
+
+  # Define the test rows, which should exclude the ref_groups
+  if (!is.null(test_groups)) {
+    df_num <- merge(df, test_groups)
+  } else {
+    df_num <- df
+    df_num <- anti_join(df_num, df_den, by = intersect(names(df_num), names(df_den)))
+  }
+
+  # Join test and denominator by their matching columns
+  merge(df_num, df_den, by = c(match_cols, "PPTESTCD"), suffixes = c("", "_den")) %>%
+    # If possible compute conversion factors for the units of test and denominator
+    mutate(
+      PPORRESU_factor = get_conversion_factor(PPORRESU_den, PPORRESU),
+      PPSTRESU_factor = if ("PPSTRESU" %in% names(.)) {
+        get_conversion_factor(PPSTRESU_den, PPSTRESU)
+      } else {
+        NULL
+      },
+      PPORRES_den = ifelse(!is.na(PPORRESU_factor), PPORRES_den * PPORRESU_factor, PPORRES_den),
+      PPSTRES_den = if ("PPSTRESU" %in% names(.)) {
+        ifelse(!is.na(PPSTRESU_factor), PPSTRESU_factor * PPSTRES_den, PPSTRES_den)
+      } else {
+        NULL
+      }
+    ) %>%
+    group_by(across(any_of(c(match_cols, group_cols, "PPTESTCD", paste0(group_cols, "_den"))))) %>%
+    unique() %>%
+    # Use mean values in case of multiple denominator rows per test
+    mutate(
+      PPORRES_den = mean(PPORRES_den, na.rm = TRUE),
+      PPSTRES_den = mean(PPSTRES_den, na.rm = TRUE),
+      n = n()
+    ) %>%
+    ungroup() %>%
+    mutate(
+      PPORRES = (PPORRES / PPORRES_den) * adjusting_factor,
+      PPSTRES = if ("PPSTRES" %in% names(.)) {
+        (PPSTRES / PPSTRES_den) * adjusting_factor
+      } else {
+        NULL
+      },
+      PPORRESU = ifelse(!is.na(PPORRESU_factor), "fraction", paste0(PPORRESU, "/", PPORRESU_den)),
+      PPSTRESU = if ("PPSTRESU" %in% names(.)) {
+        ifelse(!is.na(PPORRESU_factor), "fraction", paste0(PPSTRESU, "/", PPSTRESU_den))
+      } else {
+        NULL
+      },
+      PPTESTCD = if (!is.null(custom_pptestcd)) {
+        custom_pptestcd
+      } else {
+        ifelse(n > 1, paste0("RA", PPTESTCD, " (mean)"), paste0("RA", PPTESTCD))
+      }
+    ) %>%
+    # Keep same foramt as the input (PKNCAresults)
+    select(any_of(names(df))) %>%
+    unique()
```

Comment:
nitpick: same as above

```suggestion
    distinct()
```